mitxgraders/formulagrader.py

```diff
@@ -31,6 +31,177 @@ def casify(thestring, case_sensitive):
+            raise InvalidInput(forbidden_msg)
+    return True
+
+def validate_only_permitted_functions_used(used_funcs, permitted_functions, case_sensitive):
+    """
+    Check that the used_funcs contains only permitted_functions
+    Arguments:
+        used_functions ({str}): set of used function names
+        permitted_functions ({str}): set of permitted function names
+        case_sensitive (bool): whether comparison is case sensitive
+    Usage
+    =====
+    Case-sensitive comparison:
+    >>> validate_only_permitted_functions_used(
+    ... set(['f', 'sin']),
+    ... set(['f', 'g', 'sin', 'cos']),
+    ... True)
+    True
+    >>> validate_only_permitted_functions_used(
+    ... set(['f', 'Sin', 'h']),
+    ... set(['f', 'g', 'sin', 'cos']),
+    ... True)
+    Traceback (most recent call last):
+    InvalidInput: Invalid Input: function(s) 'h', 'Sin' not permitted in answer
+    """
+    casified_permitted = {casify(f, case_sensitive) for f in permitted_functions}
+    used_not_permitted = [f for f in used_funcs
+                          if casify(f, case_sensitive) not in casified_permitted]
+    if used_not_permitted != []:
+        func_names = ", ".join(["'{f}'".format(f=f) for f in used_not_permitted])
+        message = "Invalid Input: function(s) {} not permitted in answer".format(func_names)
+        raise InvalidInput(message)
+    return True
+
+def get_permitted_functions(whitelist, blacklist, always_allowed):
+    """
+    Constructs a set of functions whose usage is permitted.
+    Arguments:
+        whitelist ([str] | [None]): List of allowed function names
+            [None] disallows all except always_allowed
+            [] uses DEFAULT_FUNCTIONS.keys()
+        blacklist ([str]): List of disallowed function names
+        always_allowed (dict): dict whose keys are always-allowed as function names
+    Usage
+    =====
+    Whitelist some functions:
+    >>> always_allowed = {'f1': None, 'f2': None} # values unimportant
+    >>> get_permitted_functions(['sin', 'cos'], [], always_allowed) == set([
+    ... 'sin', 'cos', 'f1', 'f2'
+    ... ])
+    True
+
+    [None] disallows all defaults:
+    >>> always_allowed = {'f1': None, 'f2': None} # values unimportant
+    >>> get_permitted_functions([None], [], always_allowed) == set([
+    ... 'f1', 'f2'
+    ... ])
+    True
+
+    Blacklist some functions:
+    >>> get_permitted_functions([], ['sin', 'cos'], always_allowed) == set(
+    ... DEFAULT_FUNCTIONS).union(['f1', 'f2']).difference(set(['sin', 'cos']))
+    True
+
+    Blacklist and whitelist cannot be simultaneously used:
+    >>> get_permitted_functions(['sin'], ['cos'], always_allowed)
+    Traceback (most recent call last):
+    ValueError: whitelist and blacklist cannot be simultaneously truthy
+    """
+    # should never trigger; Grader's config validation should raise an error first
+    if whitelist and blacklist:
+        raise ValueError('whitelist and blacklist cannot be simultaneously truthy')
```

Comment:
Apparently this is triggering in some test, else coverage would not be 100%?

Reply:
That's only triggered by the doctest a few lines above. I'll edit the comment to make that clearer, though.

Reply:
See #40. A new comment here, an update to changelog, and a fresh zipfile are only edits since your last review.